k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapResponseParser.java

```diff
@@ -441,7 +441,8 @@ private String readStringUntil(char end) throws IOException {
             }
         }
 
-        throw new IOException("readStringUntil(): end of stream reached");
+        throw new IOException("readStringUntil(): end of stream reached. Read: "
+                + sb.toString() + " while waiting for '"+end+"'");
```

Comment:
Code style: operator should be at the end of the line; missing spaces around last two `+` operators.

One possible value of `end` is `\r`. We probably want to output the ASCII code rather than the character itself when `end` is less than 32.

```java
        …
        throw new IOException("readStringUntil(): end of stream reached. " +
                "Read: \"" + sb.toString() + "\" while waiting for " + formatChar(end));
    }

    private String formatChar(char value) {
        return value < 32 ? "[" + Integer.toString(value) + "]" : "'" + value + "'";
    }
```